doc/TabletRouting.md

```diff
@@ -0,0 +1,194 @@
+a keyspace, shard and tablet type. The queries are then passed to the Gateway
+implementation inside vtgate, to route them to the right tablet.
+
+There are two implementations of the Gateway interface:
+
+* discoveryGateway: It combines a set of TopologyWatchers (described in the
+  discovery section, one per cell) as a source of tablets, a HealthCheck module
+  to watch their health, and a TabletStatsCache to collect all the health
+  information. Based on this data, it can find the best tablet to use.
+* l2VTGateGateway: It keeps a map of l2vtgate processes to send queries to. See
+  next section for more details.
+  
+## l2vtgate
+
+As we started increasing the number of tablets in a cell, it became clear that a
+bottleneck of the system was going to be how many tablets a single vtgate is
+connecting to. Since vtgate maintains a streaming health check connection per
+tablet, the number of these connections can grow to large numbers. It is common
+for vtgate to watch tablets in other cells, to be able to find the master
+tablet.
+
+So l2vtgate came to exist, based on very similar concepts and interfaces:
+
+* l2vtgate is an extra hop between a vtgate pool and tablets.
+* A l2vtgate pool connects to a subset of tablets, therefore it can have a
+  reasonable number of streaming health connections. Externally, it exposes the
+  QueryService RPC interface (that has the Target for the query, keyspace /
+  shard / tablet type). Internally, it uses a discoveryGateway, as usual.
+* vtgate connects to l2vtgate pools (using the l2VTGateGateway instead of the
+  discoveryGateway). It has a map of which keyspace / shard / tablet type needs
+  to go to wich l2vtgate pool. At this point, vtgate doesn't maintain any health
+  information about the tablets, it lets l2vtgate handle it.
+
+Note l2vtgate is not an ideal solution as it is now. For instance, if there are
+two cells, and the master for a shard can be in either, l2vtgate still has to
+watch the tablets in both cells, to know where the master is. Ideally, we'd want
+l2vtgate to be collocated with the tablets in a given cell, and not go
+cross-cell.
+
+# Extensions, work in progress
+
+## Regions, cross-cell targeting
+
+At first, the discoveryGateway was routing queries the following way:
+
+* For master queries, just find the master tablet, whichever cell it's in, and
+  send the query to it. This was meant to handle cross-cell master fail-over.
+* For non-master queries, just route to the local cell vtgate is in.
+
+This turned out to be a bit limiting, as if no local tablet was available in the
+cell, the queries would just fail. We added the concept of Region, and if
+tablets in the same Region are available to serve, they can be used instead.
+This fix however is not entirely correct, as it uses the SrvKeyspace of the
+current cell, instead of the SrvKeyspace of the other cell. Both ServedFrom and
+the shard partition may be different in a different cell.
```

Comment:
The master case is actually a bit different, since there is only one. We always migrate all cells at once when migrating the master, as it doesn't make sense to have one master one way for one cell, and another way for another cell. The old / new masters are only configured one way, and blacklist all queries for what is not there, there is only one choice.

For replica and rdonly though, I kinda agree it doesn't matter much to use the wrong shard map when going cross-cell. Unless:

* there is no tablet of one kind in the other cell. Like you have both old and new tablets in the current cell, and use the new tablets, but the other cell only has old tablets.
* the other case we ran into was no tablets for a keyspace in the local cell. There was no SrvKeyspace at all. So the routing failed much earlier.

Reply:
I see -- so the problem comes up in the case where we migrate replicas differently in different cells.

Also the case of no SrvKeyspace in the current cell if there are no tablets for a keyspace is definitely an issue and is worth calling out explicitly, thanks for reminding me. From my standpoint though, that _seems_ like something we should be able to handle externally via vtctl since it is only an issue when setting up the cell for the first time or adding a new keyspace, right?